### PR TITLE
fix(native): require exact Cranelift callable slots

### PR DESCRIPTION
**Pull request by Kroisse**
## Summary
- require exact TrunkIR type equality for native callable arguments and results
- lower boxed primitive results to `core.ptr` before they fill native pointer slots
- document the contract and keep the closure regression focused on the production validator

## Validation
- `cargo clippy --workspace --all-targets -- -D warnings`
- `cargo nextest run --workspace` (2118 passed, 2 skipped)
- Negative control: restoring the old boxed `tribute_rt.anyref` result fails with the expected `clif.call_indirect` argument #2 mismatch; the production lowering was restored afterward.

## Scope
Native #960 only. This preserves the existing `core.nil` zero-width and CPS contracts. It introduces no dialect, operation, pipeline, or ownership redesign, and intentionally leaves #961 and #825 out of scope.

## Summary by CodeRabbit

- **New Features**
  - Native function signatures support zero or multiple ordered results, including `nil` values.
  - Direct and indirect native calls preserve result ordering and enforce exact type contracts.
  - Native lowering consistently uses dedicated CLIF function signatures.

- **Bug Fixes**
  - Improved validation detects malformed signatures and mismatched call, return, and pointer types.
  - Boxed native values now use consistent pointer representations.

- **Documentation**
  - Added and updated guidance for native signatures, result handling, and type contracts.

### CODE REVIEW

**Comment by Kroisse** [comment]
@coderabbitai review

File: crates/trunk-ir-cranelift-backend/src/passes/func_to_clif.rs [hunk 1]
```diff
@@ -248,45 +381,32 @@ impl RewritePattern for FuncCallIndirectPattern {
...
-                .map(|&value| ctx.value_ty(value))
-                .collect();
-            let result = result_types
-                .first()
-                .copied()
-                .unwrap_or_else(|| core::nil(ctx).as_type_ref());
-            func::func_sig(ctx, param_types, [result]).as_type_ref()
+        let Some(signature) = call.exact_signature(ctx) else {
+            return false;
+        };
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/trunk-ir-cranelift-backend/src/passes/func_to_clif.rs:386]: Addressed the nil-result mismatch in 35934716d8bd4b59cc9fe16c9905d088a29340ed. Lowering and validation now accept the exact logical result list or its defined zero-width nil projection. A live core.nil SSA result is preserved; we do not replace the logical [core.nil] contract with []. Added regressions for a live nil result through lowering and native emission, rejection of a [] signature paired with a nil SSA result, and rejection of a missing exact signature.

We are intentionally retaining the exact-signature requirement rather than reconstructing a signature from erased pointers or operand/result shapes. The closure-lowering path attaches Some(signature) before this boundary; an earlier constructor using None alone does not establish that the final native call lacks its contract. Any demonstrated routed producer that drops the contract should be fixed at that producer, not by adding a backend inference fallback.

File: crates/trunk-ir-cranelift-backend/src/translate.rs [hunk 2]
```diff
@@ -813,4 +872,49 @@ mod tests {
...
+            std::process::Command::new("cc")
+                .arg(&object_path)
+                .arg(&runtime_shim)
+                .arg("-o")
+                .arg(&executable)
+                .status()
+                .unwrap()
+                .success()
+        );
+        let status = std::process::Command::new(&executable).status().unwrap();
```
[the remark this answers is not in the document]
Comment (Kroisse) [reply crates/trunk-ir-cranelift-backend/src/translate.rs:892]: Addressed the opaque launch failure in 35934716d8bd4b59cc9fe16c9905d088a29340ed with an explicit diagnostic that cc is required to link the native execution witness. We are not adopting the skip-on-missing-cc suggestion: this test must execute the generated binary to prove zero/ordered-multiple-result ABI behavior, and an early return would report success without that acceptance evidence. Linking and execution assertions remain mandatory.

The same commit replaces the timestamp-based TestTempDir helper with tempfile::TempDir, addressing the temporary-directory isolation nit as well.
